hastory-cli/src/Hastory/Cli/OptParse.hs

```diff
@@ -1,62 +1,128 @@
+  Instructions <$> getDispatch <*> getSettings
   where
-    d =
+    getDispatch = pure dispatch
+    dispatch =
       case cmd of
-        CommandGather -> DispatchGather
-        CommandGenGatherWrapperScript -> DispatchGenGatherWrapperScript
-        CommandListRecentDirs ListRecentDirArgs {..} ->
+        CommandGather _ -> DispatchGather GatherSettings
+        CommandGenGatherWrapperScript _ ->
+          DispatchGenGatherWrapperScript
+            GenGatherWrapperScriptSettings
+              {genGatherWrapperScriptSetRemoteInfo = mbRemoteStorageClientInfo}
+        CommandListRecentDirs ListRecentDirFlags {..} ->
           DispatchListRecentDirs
-            ListRecentDirSets {lrdSetBypassCache = fromMaybe False lrdArgBypassCache}
-        CommandChangeDir i -> DispatchChangeDir i
-        CommandGenChangeWrapperScript -> DispatchGenChangeWrapperScript
-        CommandSuggestAlias -> DispatchSuggestAlias
-    sets = do
+            ListRecentDirSettings {lrdSetBypassCache = fromMaybe False lrdArgBypassCache}
+        CommandChangeDir ChangeDirFlags {..} ->
+          DispatchChangeDir ChangeDirSettings {changeDirSetIdx = changeDirFlagsIdx}
+        CommandGenChangeWrapperScript _ ->
+          DispatchGenChangeWrapperScript GenChangeWrapperScriptSettings
+        CommandSuggestAlias _ -> DispatchSuggestAlias SuggestAliasSettings
+    getSettings = do
       home <- getHomeDir
       cacheDir <-
-        case flagCacheDir of
+        case flagCacheDir <|> envCacheDir <|> configCacheDir of
           Nothing -> resolveDir home ".hastory"
-          Just fcd -> resolveDir' fcd
-      let mbBaseUrl = parseBaseUrl . T.unpack =<< flagStorageServer
-      let mbRemoteStorageClientInfo =
-            RemoteStorageClientInfo <$> mbBaseUrl <*> flagStorageUsername <*> flagStoragePassword
+          Just cd -> resolveDir' cd
       pure Settings {setCacheDir = cacheDir, remoteStorageClientInfo = mbRemoteStorageClientInfo}
-
-getConfiguration :: Command -> Flags -> IO Configuration
-getConfiguration _ _ = pure Configuration
+    mbRemoteStorageClientInfo =
+      RemoteStorageClientInfo <$> (flagStorageServer <|> envStorageServer <|> configStorageServer) <*>
+      (flagStorageUsername <|> envStorageUsername <|> configStorageUsername) <*>
+      (flagStoragePassword <|> envStoragePassword <|> configStoragePassword)
+
+getConfiguration :: Path Abs File -> Flags -> Environment -> IO Configuration
+getConfiguration defaultConfigFile Flags {..} Environment {..} =
+  maybe useDefaultConfig useProvideConfigFile (flagConfigFile <|> envConfigFile)
+  where
+    useProvideConfigFile filePath = resolveFile' filePath >>= readConfigFile >>= displayingSchema
+    useDefaultConfig = do
+      fileExists <- doesFileExist defaultConfigFile
+      if fileExists
+        then readConfigFile defaultConfigFile >>= displayingSchema
+        else pure defaultConfiguration
+    displayingSchema :: Maybe Configuration -> IO Configuration
```

Comment:
using http://hackage.haskell.org/package/yamlparse-applicative-0.1.0.1/docs/YamlParse-Applicative-IO.html#v:readConfigFile
does this all for you, the failing if something goes wrong, the rendering the error etc.
You should not have a default configuration by the way, just return a `Maybe Configuration` from `getConfiguration`.

Reply:
Thanks Syd. I've removed the idea of a default configuration.